### PR TITLE
Information about unwrapping single case unions

### PR DESCRIPTION
**Pull request by realvictorprm**
Hello fellows,
I propose to add more information about unwrapping unions especially how to unwrap single case unions.
I think this is important to know because this small details makes unions even more attractive to use.

Form and details of this can be discussed.

cc @cartermp

### CODE REVIEW

File: docs/fsharp/language-reference/discriminated-unions.md [hunk 1]
```diff
@@ -81,6 +81,24 @@ let getShapeHeight shape =
 
 Normally, the case identifiers can be used without qualifying them with the name of the union. If you want the name to always be qualified with the name of the union, you can apply the [RequireQualifiedAccess](https://msdn.microsoft.com/library/8b9b6ade-0471-4413-ac5d-638cd0de5f15) attribute to the union type definition.
 
+### Unwrapping DU's
```
Comment (cartermp) [inline docs/fsharp/language-reference/discriminated-unions.md]: Nitpick: DU's ---> Discriminated Unions

File: docs/fsharp/language-reference/discriminated-unions.md [hunk 2]
```diff
@@ -81,6 +81,24 @@ let getShapeHeight shape =
...
+
+Single case unions like this can be easily unwrapped without explicit matching.
```
Comment (cartermp) [inline docs/fsharp/language-reference/discriminated-unions.md]: It's actually a form of pattern matching to unwrap like that.  I think you're saying that it's easy to unwrap single values without needing to use the `match` or `function` expressions, which is true.

Comment (cartermp) [inline docs/fsharp/language-reference/discriminated-unions.md]: Also add a sentence that leads into the example, e.g. "The following example demonstrates unwrapping a single-case discriminated union:".

File: docs/fsharp/language-reference/discriminated-unions.md [hunk 3]
```diff
@@ -81,6 +81,24 @@ let getShapeHeight shape =
...
+
+```fsharp
+type ShaderProgram = | ShaderProgram of id:int
+```
+
+Where unwrapping follows this pattern,
```
Comment (cartermp) [inline docs/fsharp/language-reference/discriminated-unions.md]: I'd get rid of this sentence given the above.

File: docs/fsharp/language-reference/discriminated-unions.md [hunk 4]
```diff
@@ -81,6 +81,24 @@ let getShapeHeight shape =
...
+```fsharp
+let ([UnionCaseName] [values]) = [UnionValue]
+```
+
+the results for our example would be:
+```fsharp
+let (ShaderProgram id) = shaderProgram
```
Comment (cartermp) [inline docs/fsharp/language-reference/discriminated-unions.md]: Move this code sample up with the `type ShaderProgram` line.

File: docs/fsharp/language-reference/discriminated-unions.md [hunk 5]
```diff
@@ -81,6 +81,24 @@ let getShapeHeight shape =
...
+
+Single case unions like this can be easily unwrapped without explicit matching.
+
+```fsharp
+type ShaderProgram = | ShaderProgram of id:int
+```
+
+Where unwrapping follows this pattern,
+```fsharp
+let ([UnionCaseName] [values]) = [UnionValue]
```
Comment (cartermp) [inline docs/fsharp/language-reference/discriminated-unions.md]: This is a good sample showing the general case.  Perhaps move it above above your specific example with the `ShaderProgram`?

**Review by cartermp — CHANGES_REQUESTED** [review]
This is really good to mention in this article!  Thanks for getting this in.

Here's a suggested flow:

1. Mention that single-case DUs are often used for domain modeling.
2. Show the generalized code sample.
3. Mention that unwrapping the values in single-case DUs are a convenience, and that you can use pattern matching to do so without explicitly using the `match` or `function` syntax.
4. Show the specific example with `ShaderProgram`.

If you could do that, then I'll happily merge this in.

**Comment by realvictorprm** [comment]
Thanks @cartermp I'll just apply the changes

**Comment by realvictorprm** [comment]
@cartermp better now?

File: docs/fsharp/language-reference/discriminated-unions.md [hunk 6]
```diff
@@ -81,6 +81,25 @@ let getShapeHeight shape =
 
 Normally, the case identifiers can be used without qualifying them with the name of the union. If you want the name to always be qualified with the name of the union, you can apply the [RequireQualifiedAccess](https://msdn.microsoft.com/library/8b9b6ade-0471-4413-ac5d-638cd0de5f15) attribute to the union type definition.
 
+### Unwrapping Discriminated Unions
+
+In F# Discriminated Unions are often used in Domain-Modelling for wrapping a single type.
+To ease usage of such single-case discriminated unions, F# provides a nice handy syntax to unwrap them fast.
```
Comment (cartermp) [inline docs/fsharp/language-reference/discriminated-unions.md]: "To ease usage of such single-case discriminated unions, F# provides a nice handy syntax to unwrap them fast."

I'd change that to this:

"It's easy to extract the underlying value via Pattern Matching as well.  You don't need to use a `match` espression for a single case:"

File: docs/fsharp/language-reference/discriminated-unions.md [hunk 7]
```diff
@@ -81,6 +81,25 @@ let getShapeHeight shape =
...
+```fsharp
+let ([UnionCaseName] [values]) = [UnionValue]
+```
+
+The following example demonstrates unwrapping a single-case discriminated union:
```
Comment (cartermp) [inline docs/fsharp/language-reference/discriminated-unions.md]: Given the above, this can change to:

"The following example demonstrates this:"

**Review by cartermp — CHANGES_REQUESTED** [review]
Two small changes that I think should be made.  After that, it's good!  :shipit:

File: docs/fsharp/language-reference/discriminated-unions.md [hunk 8]
```diff
@@ -81,6 +81,24 @@ let getShapeHeight shape =
 
 Normally, the case identifiers can be used without qualifying them with the name of the union. If you want the name to always be qualified with the name of the union, you can apply the [RequireQualifiedAccess](https://msdn.microsoft.com/library/8b9b6ade-0471-4413-ac5d-638cd0de5f15) attribute to the union type definition.
 
+### Unwrapping Discriminated Unions
+
+In F# Discriminated Unions are often used in Domain-Modelling for wrapping a single type. It's easy to extract the underlying value via Pattern Matching as well. You don't need to use a match espression for a single case:
```
Comment (cartermp) [inline docs/fsharp/language-reference/discriminated-unions.md]: Typo: espression

File: docs/fsharp/language-reference/discriminated-unions.md [hunk 9]
```diff
@@ -81,6 +81,24 @@ let getShapeHeight shape =
 
 Normally, the case identifiers can be used without qualifying them with the name of the union. If you want the name to always be qualified with the name of the union, you can apply the [RequireQualifiedAccess](https://msdn.microsoft.com/library/8b9b6ade-0471-4413-ac5d-638cd0de5f15) attribute to the union type definition.
 
+### Unwrapping Discriminated Unions
+
+In F# Discriminated Unions are often used in Domain-Modelling for wrapping a single type. It's easy to extract the underlying value via Pattern Matching as well. You don't need to use a match expression for a single case:
```
Comment (cartermp) [inline docs/fsharp/language-reference/discriminated-unions.md]: "Domain-Modelling" --> "domain modeling"

**Review by cartermp — APPROVED** [review]

**Comment by cartermp** [comment]
Thanks @realvictorprm!  Will merge when it's green.

**Comment by realvictorprm** [comment]
You can merge @cartermp :)

**Comment by realvictorprm** [comment]
Thanks for the cooperation!

Am 16.08.2017 12:33 AM schrieb "Phillip Carter" <notifications@github.com>:

**Pull request merged**